### PR TITLE
119 repository for save party

### PR DESCRIPTION
**Pull request by kmaslowiec**
The branch's issue will be resolved after PR is approved, as it will close this one and create a new one without the history. 
Once approved, it will be merged with [119/save-party-for-merge-with-dev](https://github.com/Android-Guild/Picover/tree/119/save-party-for-merge-with-dev)

### CODE REVIEW

**Comment by MarekMacko** [comment]
@kmaslowiec Please extract migration of the `id` type to a separate PR

File: app/src/main/kotlin/com/intive/picover/parties/repository/PartiesRepository.kt [hunk 1]
```diff
@@ -20,7 +21,12 @@ class PartiesRepository @Inject constructor(
 			}
 		}
 
-	fun partyById(id: Int) =
+	suspend fun addParty(partyRemote: PartyRemote) =
+		partiesReference.push().key?.let {
+			partiesReference.child(it).setValue(partyRemote.copy(id = it)).await()
```
Comment (kmaslowiec) [inline app/src/main/kotlin/com/intive/picover/parties/repository/PartiesRepository.kt]: > Firebase clients provide a push() function that generates a unique key for each new child. For more please see: [Firebase Push().key](https://firebase.google.com/docs/database/admin/save-data#section-push)

The idea is to use Firebase's auto-generated key, assign it to the model, and use it to search for it. As the key is a String that is why the model has been changed to String.

File: app/src/main/kotlin/com/intive/picover/parties/model/Party.kt [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
-		title = title,
-		description = description,
-	)
+fun PartyRemote.toUI() = Party(
+	id = id,
+	title = title,
+	description = description,
+)
+
+fun Party.toRemote() = PartyRemote(
```
Comment (marcinWisnia) [inline app/src/main/kotlin/com/intive/picover/parties/model/Party.kt]: Do we have tests for this ?

Comment (marcinWisnia) [inline app/src/main/kotlin/com/intive/picover/parties/model/Party.kt]: I would keep this function in the PartyRemote file

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt [hunk 3]
```diff
@@ -77,8 +85,49 @@ class PartiesRepositoryTest : ShouldSpec(
...
 
 			shouldThrowExactly<NoSuchElementException> {
-				tested.partyById(1).first()
+				tested.partyById("1").first()
 			}
 		}
+
+		should("add party with correct id") {
+			val newKey = "newKey"
+			val partyRemote = PartyRemote(
```
Comment (marcinWisnia) [inline app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: I think we can use mocked values of partyRemote and modifiedPartyRemote for this test :)

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt [hunk 4]
```diff
@@ -77,8 +85,49 @@ class PartiesRepositoryTest : ShouldSpec(
...
+
+			tested.addParty(partyRemote)
+
+			coVerify {
+				childDatabaseReference.setValue(modifiedPartyRemote).await()
+			}
+		}
+
+		should("throw an exception when setValue fails") {
+			val partyRemote = PartyRemote(id = "id", title = "title", description = "desc")
```
Comment (marcinWisnia) [inline app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: here we can also use mock

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt [hunk 5]
```diff
@@ -77,8 +85,49 @@ class PartiesRepositoryTest : ShouldSpec(
...
+			coEvery { databaseReference.setValue(any<PartyRemote>()) } throws Exception("error")
```
Comment (marcinWisnia) [inline app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: here we can probably inline mocking for push and set value :)

File: app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt [hunk 6]
```diff
@@ -41,7 +42,16 @@ fun AddPartyBottomSheet(
 			onTitleChange = { viewModel.updateTitle(it) },
 			description = state.description,
 			onDescriptionChange = { viewModel.updateDescription(it) },
-			onSaveButtonClick = { navController.popBackStack() },
+			onSaveButtonClick = {
+				viewModel.addParty(
```
Comment (marcinWisnia) [inline app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt]: shouldn't this be an event ?

**Review by marcinWisnia — CHANGES_REQUESTED** [review]

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 3 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: @marcinWisnia Mocking the objects gives:` io.mockk.MockKException: no answer found for PartyRemote(#37).copy(newKey, null, null) `

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 4 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: @marcinWisnia Mocking the object gives: `expected:<"An exception"> but was:<"no answer found for PartyRemote(#47).copy(newKey, null, null)`

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 5 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: Test has been changed.

File: app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt (hunk 6 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt]: @marcinWisnia I am thinking of sending a PR for Repo now, and taking care of ViewModel within next one. Would that be ok?

File: app/src/main/kotlin/com/intive/picover/parties/model/Party.kt (hunk 2 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/parties/model/Party.kt]: Test added

Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/parties/model/Party.kt]: The function moved to PartyRemote.

File: app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt (hunk 6 above)
Comment (marcinWisnia) [reply app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt]: I am fully ok with splitting this into:
- repo
- VM 
parts :)

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 4 above)
Comment (marcinWisnia) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: ... so the copy method could be mocked as well :)

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 3 above)
Comment (marcinWisnia) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: ...so the copy method could be mocked as well :) it is better to mock this in my opinion, as i can imagine that a PartyRemote data class could have for example 10 or 15 fields. 

In such a case, it is just easier to use mocks.

**Review by marcinWisnia — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt (hunk 6 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/parties/view/AddPartyContent.kt]: Ok, in that case this and the other PR for 119 are going to be merged to [119/save-party-for-merge-with-dev](https://github.com/Android-Guild/Picover/tree/119/save-party-for-merge-with-dev)

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 3 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: Updated

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 4 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: Updated

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt [hunk 7]
```diff
@@ -80,5 +93,31 @@ class PartiesRepositoryTest : ShouldSpec(
...
+			result.shouldBeSuccess()
+		}
+
+		should("throw an exception when setValue fails") {
+			val errorMessage = "An exception"
+			every { databaseReference.push() } throws Exception(errorMessage)
+
+			val result = tested.addParty(partyRemote)
+
+			result.shouldBeFailure()
```
Comment (marcinWisnia) [inline app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: can we use assertSoftly here ?  @kmaslowiec

File: app/src/main/kotlin/com/intive/picover/parties/repository/PartiesRepository.kt [hunk 8]
```diff
@@ -20,6 +21,16 @@ class PartiesRepository @Inject constructor(
 			}
 		}
 
+	suspend fun addParty(partyRemote: PartyRemote) =
+		try {
```
Comment (bartek977) [inline app/src/main/kotlin/com/intive/picover/parties/repository/PartiesRepository.kt]: `runCatching` adds `Result` wrapper out of the box :)

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 7 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: @marcinWisnia Good one, done, thank you.

File: app/src/main/kotlin/com/intive/picover/parties/repository/PartiesRepository.kt (hunk 8 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/parties/repository/PartiesRepository.kt]: @bartek977 Really good point. thank you. Done.

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt [hunk 9]
```diff
@@ -80,5 +94,35 @@ class PartiesRepositoryTest : ShouldSpec(
...
+			}
+		}
+
+		should("throw an exception when setValue fails") {
+			val errorMessage = "An exception"
+			every { databaseReference.push() } throws Exception(errorMessage)
+
+			val result = tested.addParty(partyRemote)
+
+			assertSoftly {
```
Comment (marcinWisnia) [inline app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: please put result in method :) check the usage of this across the project :)

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt [hunk 10]
```diff
@@ -80,5 +94,35 @@ class PartiesRepositoryTest : ShouldSpec(
...
+			val childDatabaseReference: DatabaseReference = mockk {
+				every { key } returns newKey
+				coEvery { setValue(newPartyRemote).await() } returns mockk()
+			}
+			every { databaseReference.push() } returns childDatabaseReference
+			every { partyRemote.copy(newKey) } returns newPartyRemote
+
+			val result = tested.addParty(partyRemote)
+
+			assertSoftly {
```
Comment (marcinWisnia) [inline app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: i would NOT use assert softly here :) please check the method docs

**Review by marcinWisnia — CHANGES_REQUESTED** [review]

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 9 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: @marcinWisnia Fixed

File: app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt (hunk 10 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/parties/repository/PartiesRepositoryTest.kt]: @marcinWisnia Fixed

**Review by marcinWisnia — APPROVED** [review]

**Review by bartek977 — APPROVED** [review]

**Pull request closed without merging**